### PR TITLE
Adding protocol fees

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- Adding protocol fee as a hook

### Drive-by changes

- None

### Related issues

V3

### Backward compatibility

Yes

### Testing

Fuzz tests

### CODE REVIEW

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    function postDispatch(bytes calldata, bytes calldata)
+        external
+        payable
+        override
+    {
+        require(msg.value >= protocolFee, "insufficient protocol fee");
+
+        uint256 refund = msg.value - protocolFee;
+        if (refund > 0) {
+            payable(msg.sender).sendValue(refund);
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/hooks/StaticProtocolFee.sol:59]: it's a bit tricky though if this isn't the last hook that charges fees in the chain

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * @notice Collects protocol fees from the contract.
+     * @param amount The amount of protocol fees to collect. If 0, collects the entire balance.
+     */
+    function collectProtocolFees(uint256 amount) external onlyOwner {
+        uint256 amountCollected = (amount == 0)
+            ? address(this).balance
+            : amount;
+        payable(msg.sender).sendValue(amountCollected);
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:80]: I would maybe add a `beneficiary` address which is distinct from `owner` and simplify to 
```suggestion
    function collectProtocolFees() external {
        payable(beneficiary).sendValue(address(this).balance);
    }
```

File: solidity/test/hooks/ProtcolFee.t.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    function setUp() public {
+        fees = new StaticProtocolFee(1e15, address(this));
+    }
+
+    function testConstructor() public {
+        assertEq(fees.protocolFee(), 1e15);
+    }
+
+    function testSetProtocolFee(uint256 fee) public {
+        fee = bound(fee, 0, fees.MAX_PROTOCOL_FEE());
```
Comment (yorhodes) [inline solidity/test/hooks/ProtcolFee.t.sol:22]: is this the same as vm.assume <= MAX_PROTOCOL_FEE

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * @title StaticProtocolFee
+ * @notice Collects a static protocol fee from the sender.
+ * @dev V3 WIP
+ */
+contract StaticProtocolFee is IPostDispatchHook, Ownable {
+    using Address for address payable;
+
+    // ============ Constants ============
+    uint256 public constant MAX_PROTOCOL_FEE = 1e16;
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:32]: maybe immutable in constructor?

File: solidity/test/hooks/ProtcolFee.t.sol [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    function testFuzz_postDispatch_inusfficientFees(
+        uint256 feeRequired,
+        uint256 feeSent
+    ) public {
+        feeRequired = bound(feeRequired, 1, fees.MAX_PROTOCOL_FEE());
+        // bound feeSent to be less than feeRequired
+        feeSent = bound(feeSent, 0, feeRequired - 1);
+        vm.deal(alice, feeSent);
```
Comment (yorhodes) [inline solidity/test/hooks/ProtcolFee.t.sol:34]: this is a bit confusing

File: solidity/test/hooks/ProtcolFee.t.sol [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        uint256 dispatchCalls,
+        uint256 feeToBeCollected
+    ) public {
+        feeRequired = bound(feeRequired, 1, fees.MAX_PROTOCOL_FEE());
+        dispatchCalls = bound(dispatchCalls, 1, 1000);
+        feeToBeCollected = bound(
+            feeToBeCollected,
+            1,
+            dispatchCalls * feeRequired
+        );
```
Comment (yorhodes) [inline solidity/test/hooks/ProtcolFee.t.sol:76]: still quite confused

File: solidity/test/hooks/ProtcolFee.t.sol [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        fees.collectProtocolFees(feeToBeCollected);
+
+        assertEq(address(this).balance, balanceBefore + feeToBeCollected);
+    }
+
+    // ============ Helper Functions ============
+
+    receive() external payable {}
```
Comment (yorhodes) [inline solidity/test/hooks/ProtcolFee.t.sol:95]: is this so `collectProtocolFees`works?

File: solidity/test/hooks/ProtcolFee.t.sol (hunk 5 above)
Comment (aroralanuk) [reply solidity/test/hooks/ProtcolFee.t.sol:34]: what's confusing here, the feeSent < feeRequired for check revert here

File: solidity/test/hooks/ProtcolFee.t.sol (hunk 6 above)
Comment (aroralanuk) [reply solidity/test/hooks/ProtcolFee.t.sol:76]: only needed if you were able to collect less total accumulated

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    constructor(
+        uint256 _maxProtocolFee,
+        uint256 _protocolFee,
+        address _beneficiary
+    ) {
+        MAX_PROTOCOL_FEE = _maxProtocolFee;
+        _setProtocolFee(_protocolFee);
+        _setBeneficiary(_beneficiary);
+        _transferOwnership(msg.sender);
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:56]: I think making owner explicit has been helpful in deployment

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            "StaticProtocolFee: insufficient protocol fee"
+        );
+
+        uint256 refund = msg.value - protocolFee;
+        if (refund > 0) {
+            require(
+                metadata.hasSenderAddress(),
+                "StaticProtocolFee: invalid metadata"
+            );
+            payable(metadata.senderAddress()).sendValue(refund);
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:80]: I thought we would use `message.senderAddress()`?

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+     * @param _protocolFee The new protocol fee.
+     */
+    function setProtocolFee(uint256 _protocolFee) external onlyOwner {
+        _setProtocolFee(_protocolFee);
+    }
+
+    /**
+     * @notice Collects protocol fees from the contract.
+     */
+    function collectProtocolFees() external onlyOwner {
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:95]: does this need to be `onlyOwner`?

File: solidity/contracts/libs/hooks/ProtocolFeeMetadata.sol [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                )
+            );
+    }
+
+    function hasSenderAddress(bytes calldata _metadata)
+        internal
+        pure
+        returns (bool)
+    {
+        return _metadata.length == 20 && senderAddress(_metadata) != address(0);
```
Comment (yorhodes) [inline solidity/contracts/libs/hooks/ProtocolFeeMetadata.sol:30]: im not sure exactly how metadata composes but this length requirement might be prohibitive

File: solidity/test/hooks/StaticProtocolFee.t.sol [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    function setUp() public {
+        fees = new StaticProtocolFee(1e16, 1e15, bob);
+    }
+
+    function testConstructor() public {
+        assertEq(fees.protocolFee(), 1e15);
+    }
+
+    function testSetProtocolFee(uint256 fee) public {
+        fee = bound(fee, 0, fees.MAX_PROTOCOL_FEE());
```
Comment (yorhodes) [inline solidity/test/hooks/StaticProtocolFee.t.sol:23]: don't we want to check the reverting condition too?

File: solidity/test/hooks/StaticProtocolFee.t.sol [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+import {StaticProtocolFee} from "../../contracts/hooks/StaticProtocolFee.sol";
+
+contract StaticProtocolFeeTest is Test {
+    StaticProtocolFee internal fees;
+
+    address internal alice = address(0x1); // alice the user
+    address internal bob = address(0x2); // bob the beneficiary
+
+    function setUp() public {
+        fees = new StaticProtocolFee(1e16, 1e15, bob);
```
Comment (yorhodes) [inline solidity/test/hooks/StaticProtocolFee.t.sol:15]: you arent checking owner conditions

File: solidity/contracts/hooks/StaticProtocolFee.sol [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        payable
+        override
+    {
+        require(
+            msg.value >= protocolFee,
+            "StaticProtocolFee: insufficient protocol fee"
+        );
+
+        uint256 refund = msg.value - protocolFee;
+        if (refund > 0) payable(message.senderAddress()).sendValue(refund);
```
Comment (yorhodes) [inline solidity/contracts/hooks/StaticProtocolFee.sol:74]: wish there was an easy way to do this branchless
do we want to support a different refund address via metadata? similar to IGP?

File: solidity/test/hooks/StaticProtocolFee.t.sol [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        fees.collectProtocolFees();
+
+        assertEq(bob.balance, balanceBefore + feeRequired * dispatchCalls);
+    }
+
+    // ============ Helper Functions ============
+
+    function _encodeTestMessage() internal view returns (bytes memory) {
+        return
+            MessageUtils.formatMessage(
```
Comment (yorhodes) [inline solidity/test/hooks/StaticProtocolFee.t.sol:134]: still using the non standard Message library I see...

**Review by yorhodes — APPROVED** [review]

File: solidity/contracts/Mailbox.sol [hunk 16]
```diff
@@ -30,6 +30,7 @@ contract Mailbox is IMailbox, Versioned, Ownable {
 
     // A monotonically increasing nonce for outbound unique message IDs.
     uint32 public nonce;
+    // The latest dispatched message ID used for auth in post-dispatch hooks.
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/Mailbox.sol]: yaa

**Pull request merged**